### PR TITLE
Issue 1477 sensitivities for solvers

### LINKED ISSUE #1477: IDAKLU solver sensitivities using IDAS
**Summary**

Extend the IDAKLU solver to return sensitivities of y wrt the parameters p

 **Motivation**

It is useful to calculate sensitivities in, for example, model fitting to data. The IDAKLU solver already using IDA in sundials, and IDAS is another part of the sundials package which can calculate sensitivities, so in theory this *should* be easily achievable.

**Additional context**

Work on incorporating sensitivities for the casadi solver is already underway in #1100

### PR DESCRIPTION
**Pull request by martinjrobins**
# Description

Adds functionality to calculate sensitivities. Specifically:
 - `pybamm.base_solver.solve` function can take a list of input parameters to calculate the sensitivities of the solution with respect to. Or it can be `True` to calculate the sensitivities for all input parameters. If this arg changes between different calls to `solve`, then the model is re-setup.
 - If `model.convert_to_format = 'casadi'`, then sensitivities will be calculated by solving the explicit forward sensitivity equations. Any solver can be used, except for the IDKLU solver which uses CVODES forward sensitivity methods, see next item
 - For `model.convert_to_format = 'casadi'` or `model.convert_to_format = 'jax'`, the IDKLU solver can use the CVODES package to calculate sensitivities.

Fixes #1477

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)

# Key checklist:

- [x] No style issues: `$ flake8`
- [x] All tests pass: `$ python run-tests.py --unit`
- [x] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: sens-test-pybamm.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import pybamm
```
Comment (tinosulzer) [inline sens-test-pybamm.py:1]: can this file be deleted or moved somewhere else?

File: sens-test.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
+import casadi
```
Comment (tinosulzer) [inline sens-test.py:1]: same here?

File: test-sensitivities.ipynb [hunk 3]
```diff
@@ -0,0 +1,604 @@
+{
```
Comment (tinosulzer) [inline test-sensitivities.ipynb:1]: should this be in teh examples folder or deleted?

File: tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_dfn.py [hunk 4]
```diff
@@ -21,6 +21,21 @@ def test_basic_processing(self):
...
+        # use Ecker parameters for nonlinear diffusion
+        param = pybamm.ParameterValues(chemistry=pybamm.parameter_sets.Ecker2015)
+        var = pybamm.standard_spatial_vars
+        var_pts = {var.x_n: 10, var.x_s: 10, var.x_p: 10, var.r_n: 5, var.r_p: 5}
+        modeltest = tests.StandardModelTest(
+            model, parameter_values=param, var_pts=var_pts
+        )
+        modeltest.test_sensitivities(
+            #'Separator thickness [m]', 2e-05,
+            'Typical current [A]', 0.15652,
```
Comment (tinosulzer) [inline tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_dfn.py:36]: can we test wrt "Current function [A]" instead of "Typical current [A]"? Typical current is a scaling parameter

File: tests/unit/test_solvers/test_casadi_algebraic_solver.py [hunk 5]
```diff
@@ -174,157 +172,6 @@ def test_solve_with_input(self):
         np.testing.assert_array_equal(solution.y, -7)
 
 
-class TestCasadiAlgebraicSolverSensitivity(unittest.TestCase):
```
Comment (tinosulzer) [inline tests/unit/test_solvers/test_casadi_algebraic_solver.py:177]: why was this deleted?

**Review by tinosulzer — COMMENTED** [review]
THanks for taking this on @martinjrobins ! Just a few comments

File: CHANGELOG.md [hunk 6]
```diff
@@ -174,12 +175,13 @@ This release adds new operators for more complex models, some basic sensitivity
 
 ## Breaking changes
 
+-   Changed sensitivity API. Removed `ProcessedSymbolicVariable`, all sensitivity now handled within the solvers and `ProcessedVariable` ()
+-  Renamed `quick_plot_vars` to `output_variables` in `Simulation` to be consistent with `QuickPlot`. Passing `quick_plot_vars` to `Simulation.plot()` has been deprecated and `output_variables` should be passed instead ([#1099](https://github.com/pybamm-team/PyBaMM/pull/1099))
```
[the remark this answers is not in the document]
Comment (martinjrobins) [reply CHANGELOG.md:179]: Yea, this was left-over from your branch, I'll tidy it up

File: sens-test-pybamm.py (hunk 1 above)
Comment (martinjrobins) [reply sens-test-pybamm.py:1]: I didn't notice this one, I'll delete it

File: sens-test.py (hunk 2 above)
Comment (martinjrobins) [reply sens-test.py:1]: and here

File: tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_dfn.py (hunk 4 above)
Comment (martinjrobins) [reply tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_dfn.py:36]: sure, I'll give "Current function" a go

File: tests/unit/test_solvers/test_casadi_algebraic_solver.py (hunk 5 above)
Comment (martinjrobins) [reply tests/unit/test_solvers/test_casadi_algebraic_solver.py:177]: that was a mistake on my part, I'll put it back in

**Review by tinosulzer — APPROVED** [review]
Looks good! Will just need to merge develop after #1596 to fix the github actions
